update doc; /regions signature; /grids signature (#4)
contrailcirrus/contrail-forecast: docs/specs/forecast-api.md

    hour interval, at all longitudes and latitudes ranging from \[-80,
    80\] (this is expected to expand and encompass the poles, with an
    update at a later time). These grids are available across several
    [aircraft classes](#aircraft-classes); only the `default` aircraft class is served until 

>     there is sufficient customer evidence to suggest needing the implementation of additional aircraft classes. 

Reviewer: ```suggestion
    there is sufficient evidence and agreement to implement additional aircraft classes. 
```

    flight levels and temporal intervals as the grid above. Regions are
    available at pre-defined `contrails` thresholds of \[1, 2, 3, 4\].

These two endpoints represent a minimal
yet complete set of capabilities for air traffic planners to explore and
prototype contrail avoidance in air traffic decision support systems.

The signature, behavior and design considerations of these two endpoints
are detailed below.

> Future work, contingent upon the implementation of multiple aircraft classes,

Reviewer: ```suggestion
If/when multiple aircraft classes are implemented, future work may include a discovery/recommendation endpoint to
```

Reviewer: (semantic suggestion that isn't necessarily better, feel free to reject)

### timestamp

> The `ts` value has the form `%Y%m%d%H`. It is a required query parameter.

Reviewer: ```suggestion
The `{ts}` value has the form `%Y%m%d%H`. It is a required query parameter.
```

### flight level

> The {fl} value represents the flight level (*hectofeet*) of the returned

Reviewer: Shouldn't the {fl} be in a code block, i.e. `{fl}`?

Author: yes; inline. I'll also scan for other similar fmt fixes

> The `{fl}` value represents the flight level (*hectofeet*) of the returned

Reviewer 2: Could you rename this to `flight_level`? This would match both the result coordinate name and our parameter/coordinate name.

Author: You mean the param name, right? If so, I think we're already good (see signature here):
https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#signature-1

The bracketed fields just represent the parameter value passed.
We're just using shorthand for brevity in the signature documentation, but if you'd prefer that the signature read:
`?flight_level={flight_level}&time={time}` etc... instead of `?flight_level={fl}&time={ts}`, I'm OK with that.

Reviewer 2: This was an oversight from my end – the short form for the placeholder is perfectly fine! Thanks for pointing it out :)

### flight level

> The fl value represents the flight level (*hectofeet*) of the returned

Reviewer: The `{fl}` value represents the flight level (*hectofeet*) of the returned

> The fl value represents the flight level (*hectofeet*) of the returned

Author: consistency fix:
```suggestion
The {fl} value represents the flight level (*hectofeet*) of the returned
```

  * latitude   (latitude) float32 3kB -80.0 -79.75 -79.5 ... 79.5 79.75 80.0
  * flight_level      (flight_level) int16 2B 300
  * time       (time) datetime64[ns] 8B 2024-07-01T12:00:00
    forecast_reference_time (time) datetime64[ns] 8B 2024-07-01T06:00:00
Data variables:
    contrails   (longitude, latitude, flight_level, time) float32 4MB ...
Attributes:
    aircraft_class: "default"
```

Note that `forecast_reference_time` non-dimension coordinate has the same definition as

> `model_run_at` in the API response header.

Reviewer: Are we still including `model_run_at`? Is this comment necessary anymore?

Author: I don't think so (and agreeing w/ you suggestion below to just include `forecast_reference_time` in the geoJSON properties).

We can strike it here (and elsewhere), thereby making the `model_*` quantities something of an internal definition/concern only.

  * latitude   (latitude) float32 3kB -80.0 -79.75 -79.5 ... 79.5 79.75 80.0
  * flight_level      (flight_level) int16 2B 300
  * time       (time) datetime64[ns] 8B 2024-07-01T12:00:00

>     forecast_reference_time (time) datetime64[ns] 8B 2024-07-01T06:00:00

Reviewer 2: A tiny question for `time` and `forecast_reference_time`: Is it intentional to remove the timezone `Z` from the otherwise ISO8601 timestamp?

Author: great question. I believe time-like data is actually stored as int64 in the netCDF -- the types here reflect the marshaling that xarray does when importing the file, so the timezone naive type here is auto-magically set by xarray.

@mlshapiro can you check me on correctness here?

Author: That does bring up a good questions as to whether we'd want to suffix `Z` on the timestamps returned in the geoJSON properties, however: https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#response-object-1

As it is now, we have consistency. Timezone naive timestamps throughout; implicit UTC.

Adding suffixes in the Propeties of the geoJSON may create mixed conventions, but add specificity to the geoJSON response, at least.

I could see arguments for either way. But I leaning towards consistency (and of course, there is always documentation).

Author: Also noting that the `timestamp` passed as a query param (`timestamp="%Y%m%d%H"`) is timezone naive, UTC implicit as well.

If we wanted to be timezone aware here... we'd probably want to go to `timestamp=<int>` unixtime or `timestamp=<iso_datetime_str>`.  I see that being more canonical in general... but we loose the self-documenting goodness of the current design (which communicates to customers that response data only exists on the hour).

Reviewer 2: Could you change the signature to `time=ISO 8601`? Renaming to `time` would match both the coordinate in your return data and our parameter/coordinate name. I'd personally prefer explicit documentation around the coarseness of times, while allowing any ISO 8601 time to be passed in. We might change the forecast to exist more often than only once per hour in the future.

Author: > We might change the forecast to exist more often than only once per hour in the future.

👍 

> Could you change the signature to time=ISO 8601? Renaming to time would match both the coordinate in your return data and our parameter/coordinate name. I'd personally prefer explicit documentation around the coarseness of times, while allowing any ISO 8601 time to be passed in.

Completely agreed on consistency of naming (`timestamp` -> `time`).

On field definition loosening to any iso8601 date(time) str..., we loose the benefits of the quantity being self-describing which is a downside , and generally loosen product-fit. But I see the motivation given the general trend here towards maintaining flexibility at the expensive of product fit.  Given that you'll be owning this long-term, I see no problem in updating to accept any iso8601 string.

If we make that change, I see two behavioral considerations:
1) what would be your preferred behavior if the provided value doesn't match an available prediction? For instance, predictions are currently only available hourly (aligned w/ frequency of meteorological forecast availability).  If a caller provides `"2024-01-02T01:12:00"`, would you prefer we snap to the nearest available forecast, or return a 404, or other?

2) do you want to support timezone handling? should the tz designator be required? If not required, then implicit UTC?

Reviewer 2: 1. Snapping to the nearest forecast would be perfect. This would allow us to be flexible in what timestamps we accept which IMO makes the API easily usable. Callers can then learn which actual forecast was provided in the response through the reference timestamp dimension.

2. No strong opinion, your proposal sounds good.

Data variables:
    contrails   (longitude, latitude, flight_level, time) float32 4MB ...
Attributes:

>     forecast_reference_time:  "2024-07-01T06:00:00Z"

Reviewer 2: qq: What's the reason for migrating `forecast_reference_time` to be a coordinate and no longer an attribute? We have a similar attribute today and would then evaluate the same change!

Author: If you want to commit to only ever serving a single time slice, then I think having it as an attr makes sense. If you want to option available to return multiple time slices, then it should move to a coord (that references `time` 1:1).

--

The `forecast_reference_time` is the time at which the _meteorological model_ (used in generating the contrails predictions) was executed for a given `time`.  So multiple time slices may have different `forecast_reference_time` values.

Possible alternatives I can think of, but that seem less favorable:
- support multiple times in query but return multiple netcdfs for a given call, keeping `forecast_reference_time` as an attr
- support multiple times in query and concat slices into a single netcdf, but have all time slices reference a single `forecast_reference_time` (this would (1) limit the valid time slice candidates that could be concat'ed into a single file, and, (2) force choosing `time` slices that share a common `forecast_reference_time` instead of the latest available `forecast_reference_time` for a given `time`.

--
FWIW, I think its a good idea to stay away from supporting multiple time slices or time ranges in a query. It would make adding a 305 response option tricky in the future (many customers we've talked to expect to use the API as part of a caching service, so a 305 would be helpful).

Author: pinning this here as it is a relevant reminder for context ⏫ 

https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#timestamp

> Given that ECMWF delivers 73 hours of forecast data every 6 hours, it is expected that a given ts will have multiple candidate URIs to serve to a client.

> If multiple candidate URIs exist for a given timestamp at the time of the client request, the API should return the URI whose outputs were generated *using the nearest forecast time*.

Reviewer 2: Thanks for elaborating, this really helped my understanding!

1) What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

2) Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

Author: > What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

TL;DR -- if we have N time slices, it is very possible that only a few slices are modified at any given time and warrant cache updates.  Sending N slices over the wire every time is sub-optimal.  Providing change detection (modification) on a per time slice basis, and enforcing retrieval on a per-time-slice basis, more strongly guarantees a client will have quality caching behavior (albeit with a bit more complexity to manage concurrency).

Longer answer:
If the contrail predictions are using ECMWF meteorological forecasts, then there will be (nominally) 12 different met forecasts for a given target `time`.  (ECMWF provides forecasts 73hrs into the future, now() inclusive, and they re-run the forecast every 6hrs, so any given hourly `time` has 73/6 overlapping forecast options).

As such, a query for a given `?time=splat` can return 12 different contrail forecasts, depending on the time of the call (default behavior, as per customer request and expected usage, being to return the latest forecast available).  This is why there is a note about the API not being idempotent. This is akin to checking for Friday's weather today, then checking for Friday's weather again tomorrow, and expecting the weather values to be different.  Noting that a future "super-user" upgrade could certainly be to accept a `?forecast_time=splat` param, which would guarantee idempotent behavior, but make client integration logic considerably more complex.

So to the question of caching logic wrt multiple time slices, a given netCDF file would have 12**N variants given N `time` slices in the netCDF.  A 305 response (or other similar change detection mechanism) would trigger whenever _any_ of the N slices is modified.  If a client is well informed wrt how the forecasts are due to update, they could get a high hit-rate of modified slides (N_mod/N_tot ~=1), but, making this assumption seems antithetical to a flexible API design (separating domain logic concerns from integration logic concerns). 

This is a bit sticky and I can see a few other ways that this could have been designed, but I won't elaborate here (this design was chosen given our opinions on how flight planners will want machine-2-machine integration without a big integration logic burden).  I'm happy to jump on a call and discuss the time/forecast_time behavior in more depth if you're interested.

> Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

Great question. I actually don't know from a netCDF consumer perspective. I'll inquire w/ our team today as they have more experience working w/ netCDF using various client-side tools. 

I _would_ expect that having the `forecast_time` as an attr, and then removing it (migrating to a coord) would be breaking.  So perhaps we keep it as a coord to start? (I don't see other substantive downside other than that `forecast_time` "feels" more like an attr than a first-class coord).

Author: following up here.
It shouldn't be breaking to add a new non-indexed coordinate to the netCDF at a future time.

It is important, however, to include `forcast_reference_time`... and if we include it in an attribute and then decide to migrate to a non-indexed coord later, it would be a breaking change to remove the _attribute_ in favor of the new non-indexed coordinate.

So if there is any possibility of wanting to return multiple `time` slices in the future, it seems reasonable to just have `forecast_reference_time` be a non-indexed coord (or, we can agree that we won't leave the back door open for aggregations on `time` and leave `forecast_reference_time` an attr... ◀️  my preference).

  "features": [
      {
        "type": "Feature",
        "properties": {
            "aircraft_class": "default",

>             "timestamp": "<%Y-%m-%dT%H:%M:%S>",

Author: probably makes sense to suffix `Z` here  for avoidance of doubt (`"<%Y-%m-%dT%H:%M:%SZ>"`), if there is a future possibility of accepting timezone aware strings in the `time` query param.

Author: this will also be renamed to `time`

            "flight_level": <int>,
            "threshold": <int>,

>             "forecast_reference_time": <%Y-%m-%dT%H:%M:%S>

Author: probably makes sense to suffix Z here for avoidance of doubt ("<%Y-%m-%dT%H:%M:%SZ>"), if there is a future possibility of accepting timezone aware strings in the time query param.

  "features": [
      {
        "type": "Feature",
        "properties": {
            "aircraft_class": "default",
            "timestamp": "<%Y-%m-%dT%H:%M:%S>",
            "flight_level": <int>,
            "threshold": <int>,

>             "model_forecast_hour": <int>,

Reviewer: Are we including any `model_...` properties anymore?

My suggestion would be to just include the `forecast_reference_time` only to be consistent with the netCDF file.  If you want to include `forecast_period` (instead of `model_forecast_hour`) this would be the integer hours between `timestamp` and `forecast_reference_time`

Or if you want to include these, then copy the definitions from up above down here. I would only include

Rather just the `forecast_reference_time` the same as in the netCDF file?

Author: I agree.

I think we can strike the `model_*` quantities (for two reasons).

1. nice to make this nomenclature an internal concern only
2. anybody concerning themselves with model run at vs. predicted at... will be well enough informed to know that the `"timestamp"` quantity here is equivalent to `"predicted_at"`, `"forecast_reference_time"` is equiv. to `"model_run_at"`, and `"model_forecast_hour"` is the diff of the two.
